docs: Added documentation for automatic TCP and UDP ports waiting

leoporoli opened the pull request:
## Description:
Added documentation for automatic TCP and UDP ports waiting
Added extra information on when users can use ports waiting or ReadyCondition or the wait instruction

## Is this change user-facing?
YES

## References (if applicable):
https://www.notion.so/kurtosistech/Automatic-TCP-Port-Waiting-17f27247ddc14cca930f5cff5627319b

https://github.com/kurtosis-tech/kurtosis/issues/192

leoporoli committed:
Specifying when to use ReadyConditions, the wait instruction or just the ports wait

leoporoli committed:
Adding the wait timeout field in the starlark PortSpec

leoporoli committed:
Merge branch 'main' into lporoli/ticket-192-part-2

leoporoli committed:
last changes

leoporoli committed:
Merge branch 'main' into lporoli/ticket-192-part-2

leoporoli committed:
Reviewing one test

leoporoli committed:
Fixed test

leoporoli committed:
Fixed TestStartosisInterpreter_ValidSimpleScriptWithInstructionPortNumberAsString

leoporoli committed:
Added wait in some tests

leoporoli committed:
handling service logs errors

leoporoli committed:
handling service logs errors

leoporoli committed:
Added internal integration tests

leoporoli committed:
Merge branch 'main' into lporoli/ticket-192-part-2

leoporoli committed:
Fixed using found

leoporoli committed:
Fixed test helpers

leoporoli committed:
Default value in service args

leoporoli committed:
Changes after PR review

leoporoli committed:
Merge branch 'main' into lporoli/ticket-475

leoporoli committed:
Merge branch 'main' into lporoli/ticket-192-part-2

leoporoli committed:
Merge branch 'main' into lporoli/ticket-192-part-2

leoporoli committed:
Merge branch 'lporoli/ticket-192-part-2' into lporoli/ticket-475

leoporoli committed:
Added documentation for wait in port spec

leoporoli committed:
Merge branch 'lporoli/ticket-192-part-2' into lporoli/ticket-475

leoporoli commented:
@leeederek you can validate only the docs change, t[here is another Pr for the code](https://github.com/kurtosis-tech/kurtosis/pull/534) but it's not merged yet because I want to merge both at the same time.

h4ck3rk3y committed:
docs: ask users to just click on "New Workspace" due to recent GitPod changes (#556)

## Description:
Kevin found the popup to be confusing; agree with the feedback. Haven't
found a workaround for it so adding more docs as duct tape while we
discuss the issue here https://github.com/gitpod-io/gitpod/issues/17215

## Is this change user facing?
YES

## References (if applicable):
Closes #555

h4ck3rk3y committed:
ci: limit the number of versions of current minor version of docs kept & historical versions kept (#552)

## Description:
Enforces limits using

HISTORICAL_VERSIONS_TO_KEEP = 5
MOST_TO_KEEP_OF_LATEST_VERSION = 5

If they're set to -1 then we keep all of the current version.

## Is this change user facing?
YES

h4ck3rk3y committed:
docs: added some tips about commands (#561)

## Description:
Earlier the way commands were run using `/bin/sh` `-c` was an internal
thing we knew and now we share it with the world.

## Is this change user facing?
YES

---------

Co-authored-by: Victor Colombo <victorcolombo@users.noreply.github.com>

leoporoli commented:
hey @leeederek don't forget this guy ;)

leeederek commented on docs/docs/starlark-reference/port-spec.md:
```diff
@@ -21,10 +21,25 @@ port_spec = PortSpec(
     # OPTIONAL (DEFAULT:"http")
     application_protocol = "http",
+    
+    # All the service's TCP and UDP ports are automatically check, Kurtosis check if the declared ports
```
```suggestion
    # Kurtosis will automatically perform a check to ensure all declared UDP and TCP ports are open and ready for traffic and connections upon startup.
```

leeederek commented on docs/docs/starlark-reference/port-spec.md:
```diff
@@ -21,10 +21,25 @@ port_spec = PortSpec(
     application_protocol = "http",
+    
+    # All the service's TCP and UDP ports are automatically check, Kurtosis check if the declared ports
+    # are open before finishing the service startup.
```
```suggestion
```

leeederek commented on docs/docs/starlark-reference/port-spec.md:
```diff
@@ -21,10 +21,25 @@ port_spec = PortSpec(
+    
+    # All the service's TCP and UDP ports are automatically check, Kurtosis check if the declared ports
+    # are open before finishing the service startup.
+    # We can set a custom wait time out value with or disable the feature or through this property
```
```suggestion
    # You may specify a custom wait timeout duration or disable the feature entirely.
```

leeederek commented on docs/docs/starlark-reference/port-spec.md:
```diff
@@ -21,10 +21,25 @@ port_spec = PortSpec(
+    # All the service's TCP and UDP ports are automatically check, Kurtosis check if the declared ports
+    # are open before finishing the service startup.
+    # We can set a custom wait time out value with or disable the feature or through this property
+    # You can specify a custom wait time out using a Duration string:
```
```suggestion
    # You may specify a custom wait timeout duration with a string:
```

leeederek commented on docs/docs/starlark-reference/port-spec.md:
```diff
@@ -21,10 +21,25 @@ port_spec = PortSpec(
+    # We can set a custom wait time out value with or disable the feature or through this property
+    # You can specify a custom wait time out using a Duration string:
+    #  wait = "2m"
+    # Or, you can disable this feature by setting the value in None:
```
```suggestion
    # Or, you can disable this feature by setting the value to None:
```

leeederek commented on docs/docs/starlark-reference/port-spec.md:
```diff
@@ -21,10 +21,25 @@ port_spec = PortSpec(
+    #  wait = "2m"
+    # Or, you can disable this feature by setting the value in None:
+    #  wait = None
+    # The feature is enabled by default with a default timeout
```
```suggestion
    # The feature is enabled by default with a default timeout of 15s
```

leeederek commented on docs/docs/starlark-reference/port-spec.md:
````diff
@@ -21,10 +21,25 @@ port_spec = PortSpec(
 ```
 The above constructor returns a `PortSpec` object that defines information about a port for use in [`add_service`](../concepts-reference/subnetworks.md).
 
+The wait field represents the timeout for the automatic ports opening check, which is the default way to check for readiness. There are others ways to check service readiness, you can use the [ServiceConfig.ReadyConditions][ready-conditions] which can be used to check if a service is ready through a REST call for instance, or you can use the [Wait][wait] instructions if you need a one-off check f ...
````
```suggestion
The `wait` field represents the timeout duration that Kurtosis will use when checking whether or not a service's declared UDP or TCP port are open and ready for traffic and connections upon startup. This is the default way to perform a readiness check using Kurtosis. However, there are other ways to perform a readiness check. Specifically, you can also use [`ServiceConfig.ReadyConditions`][ready-conditions] to check if a service is ready with a a REST call, or you can use the [Wait][wait] instruction if you need to perform a one-off readiness check after service startup.
```

leeederek commented on docs/docs/starlark-reference/ready-condition.md:
```diff
@@ -5,6 +5,8 @@ sidebar_label: ReadyCondition
 
 The `ReadyCondition` can be used to execute a readiness check after a service is started to confirm that it is ready to receive connections and traffic 
```
```suggestion
The `ReadyCondition` can be used to execute a readiness check after a service is started to confirm that it is ready to receive connections and traffic.
```

leeederek commented on docs/docs/starlark-reference/ready-condition.md:
```diff
@@ -43,4 +45,42 @@ ready_conditions = ReadyCondition(
+
 <!--------------- ONLY LINKS BELOW THIS POINT ---------------------->
+
+[service-config]: ./service-config.md
```
```suggestion
[service-config]: ./service-config.md
[port-spec]: ./port-spec.md
```

leeederek commented on docs/docs/starlark-reference/ready-condition.md:
```diff
@@ -5,6 +5,8 @@ sidebar_label: ReadyCondition
 
 The `ReadyCondition` can be used to execute a readiness check after a service is started to confirm that it is ready to receive connections and traffic 
 
+It is flexible, and you can configure it in order to check service readiness in many ways, but if you just want to check if your service's ports are open (to consider it ready) and these can receive incoming calls, we suggest to use the `wait` field in the [ServiceConfig][service-config] type. And if you need one-off wait not attached to the [ServiceConfig][service-config] you suggest you to use  ...
```
Re-worded a bit, took out the advice for using plan.wait because readyCondition has the same functionality. Let me know what you think! 

It's our intention to replace `wait` with `readyCondition` entirely for more complex wait logic, right?
```suggestion
As you will see below, using `ReadyCondition` is a flexible and highly configurable way to define a readiness check for a given service. However, if all you need is a check upon service startup for whether or not a port is open and ready for traffic, then we recommend relying on the default `wait` field in the [PortSpec constructor][port-spec] (as part of the [ServiceConfig][service-config] type).
```

leeederek reviewed (approved):
I can't approve the PR for the code changes, but the docs are good if you accept my suggestions 👍🏻

victorcolombo committed:
docs: Address `args` change from `struct` to `dict` in quickstart (#564)

## Description:
<!-- Describe this change, how it works, and the motivation behind it.
-->

## Is this change user facing?
YES
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
<!-- Add relevant Github Issues, Discord threads, or other helpful
information. -->

leoporoli committed:
Update docs/docs/starlark-reference/port-spec.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/port-spec.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/port-spec.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/port-spec.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/port-spec.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/ready-condition.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/port-spec.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/ready-condition.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/ready-condition.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Update docs/docs/starlark-reference/port-spec.md

Co-authored-by: Derek Lee <103802618+leeederek@users.noreply.github.com>

leoporoli committed:
Merge branch 'main' into lporoli/ticket-475
